compiler/ballerina-lang/src/main/java/io/ballerina/projects/Package.java

```diff
@@ -412,8 +415,22 @@ private Package createNewPackage() {
             PackageContext newPackageContext = new PackageContext(this.project, this.packageId, this.packageManifest,
                     this.ballerinaTomlContext, this.dependenciesTomlContext, this.cloudTomlContext,
                     this.compilerPluginTomlContext, this.packageMdContext,  this.compilationOptions,
-                    this.moduleContextMap, this.pkgDependencyGraph);
+                    this.moduleContextMap, DependencyGraph.emptyGraph());
             this.project.setCurrentPackage(new Package(newPackageContext, this.project));
+
+            List<ResolvedPackageDependency> newSortedList = this.project.currentPackage().getResolution()
+                    .dependencyGraph().toTopologicallySortedList();
+            List<ResolvedPackageDependency> oldSortedList = this.dependencyGraph.toTopologicallySortedList();
+
+            for (int i = 0; i < newSortedList.size(); ++i) {
```

Comment:
We can compare the sizes before comparing each item. will save bit of time when a new dependency is added. But ideally when a new dependency is added we do not need to clear the cache.

Reply:
+1. We can improve this logic a bit more. 

So we have two dependency graphs here. Old graph O and new graph N. Can we figure out the number of node additions and the number of node deletions that we need to apply to graph O to get graph N? We don't worry about the structure of the graph; we just need to additions and deletions. 

If there are deletions, we clear the cache, otherwise keep the cache as it is.

Reply:
In the case of Dependencies.toml edit, there is only a version change in nodes. The number of nodes will stay consistent since the imports are resolved from the source files and only the version is captured from the Dependencies.toml. This improvement will be applicable for document edits (since there can be a change in import statements) but can we apply this for this case?

Reply:
Yes, what we need is to figure out deletions. Deletions should tell you whether you need to clear the cache or not.

Reply:
Addressed in 1c11376